modules/distributor/forwarder.go

```diff
@@ -0,0 +1,248 @@
+	traces []*rebatchedTrace
+}
+
+// forwarder queues up traces to be sent to the metrics-generators
+type forwarder struct {
+	services.Service
+
+	// per-tenant queue managers
+	queueManagers map[string]*queueManager
+	mutex         sync.Mutex
+
+	forwardFunc forwardFunc
+
+	o        *overrides.Overrides
+	shutdown chan interface{}
+}
+
+func newForwarder(fn forwardFunc, o *overrides.Overrides) *forwarder {
+	rf := &forwarder{
+		queueManagers: make(map[string]*queueManager),
+		mutex:         sync.Mutex{},
+		forwardFunc:   fn,
+		o:             o,
+		shutdown:      make(chan interface{}),
+	}
+
+	rf.Service = services.NewBasicService(rf.start, rf.running, rf.stop)
+
+	return rf
+}
+
+// SendTraces queues up traces to be sent to the metrics-generators
+func (rf *forwarder) SendTraces(ctx context.Context, userID string, keys []uint32, traces []*rebatchedTrace) {
+	rf.mutex.Lock()
+	defer rf.mutex.Unlock()
+
+	if _, ok := rf.queueManagers[userID]; !ok {
+		rf.queueManagers[userID] = newQueueManager(userID, rf.forwardFunc, rf.o)
+	}
+
+	err := rf.queueManagers[userID].push(ctx, &pushRingRequest{keys: keys, traces: traces})
+	if err != nil {
+		metricForwarderDroppedPushes.WithLabelValues(userID).Inc()
+	}
+}
+
+// watchOverrides watches the overrides for changes
+// and updates the queueManagers accordingly
+func (rf *forwarder) watchOverrides() {
+	for {
+		select {
+		case <-time.After(time.Second * 5):
+			rf.mutex.Lock()
+			for _, tm := range rf.queueManagers {
+				tm.watchOverrides(rf.o)
+			}
+			rf.mutex.Unlock()
+		case <-rf.shutdown:
+			return
+		}
+	}
+}
+
+func (rf *forwarder) start(_ context.Context) error {
+	go rf.watchOverrides()
+
+	return nil
+
+}
+
+func (rf *forwarder) running(ctx context.Context) error {
+	<-ctx.Done()
+	return nil
+}
+
+func (rf *forwarder) stop(_ error) error {
+	close(rf.shutdown)
+	var errs []error
+	for _, tm := range rf.queueManagers {
+		if err := tm.shutdown(); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return multierr.Combine(errs...)
+}
+
+// queueManager manages a single tenant's queue
+type queueManager struct {
+	mutex sync.Mutex
+	wg    sync.WaitGroup
+
+	tenantID       string
+	numWorkers     int
+	queueSize      int
+	reqChan        chan *pushRingRequest
+	fn             forwardFunc
+	workersCloseCh chan struct{}
+
+	shutdownCh chan struct{}
+}
+
+func newQueueManager(tenantID string, fn forwardFunc, o *overrides.Overrides) *queueManager {
+	m := &queueManager{
+		mutex:      sync.Mutex{},
+		tenantID:   tenantID,
+		fn:         fn,
+		shutdownCh: make(chan struct{}),
+	}
+
+	m.watchOverrides(o)
+	return m
+}
+
+// push a trace to the queue
+// if the queue is full, the trace is dropped
+func (m *queueManager) push(ctx context.Context, req *pushRingRequest) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	select {
+	case m.reqChan <- req:
+		// TODO: Record some metric
+	case <-ctx.Done():
+		return fmt.Errorf("failed to push traces to tenant %s queue", m.tenantID)
```

Comment:
do we need a default in case the queue is full? i'm not sure it makes sense to check if the context is done here or not. this will execute so quickly that i don't think we care.

Reply:
Addressed in 4b5aeb3ef729ddc22dfba07139911ff75cbc7b46